TokenFactory.mon

```diff
@@ -1,194 +1,209 @@
-			}
-			//	Always respond with the cached value even if it is invalid.
-			//	A new TokenUpdate notification will be sent on every future refresh
-			send TokenRegistrationResponse(req.reqId, tokenCache[req.eventType].token) to 
-							TokenRegistrationResponse.SUBSCRIBE_CHANNEL;
-		}
-		
-		on all RegistrationCancellationReq() as req {
-			string cacheKey := req.eventType;
-			if not tokenCache.hasKey(cacheKey) {
-				log "Trying to unregisterInterest with an invalid TokenRegistrar."
-					+ " Use TokenRegistrar.registerInterest() to get a valid registrar" at WARN;
-				return;
-			}
-			
-			CacheEntry cachedEntry := tokenCache[cacheKey];
-			cachedEntry.consumers := cachedEntry.consumers - 1;
-			if (cachedEntry.consumers = 0) {
-				send StopEventListener(req.eventType) to context.current();
-				tokenCache.remove(req.eventType);
-			}
-		}
-	}
-	
-	/** 
-	*	Starts an expiration timer on a cache key.
-	*/
-	action startExpiryTimer(string key) {
-		//	we can configure an additional wait to log a early warning, 
-		//	trigger a refresh-token subroutine or notify the person responsible via SMS/EMAIL
-		
-		on wait(tokenCache[key].expiration - currentTime) and not StopEventListener(type=key){
-			// check if the token was refreshed during this wait
-			CacheEntry cachedEntry := tokenCache[key];
-			if (currentTime >= cachedEntry.expiration) {
-				log "Token with " + key + " as key has expired and no longer valid." at WARN;
-				// set token value as empty which implies invalid token
-				cachedEntry.token := "";
-				
-				// Notify the consumers of this token
-				send TokenUpdate(key, cachedEntry.token) to TokenRegistrationResponse.SUBSCRIBE_CHANNEL;
-			}
-		}
-	}
-	
-	/** 
-	*	Updates the caches on any token refreshes received as events and notifies 
-	*	the consumers by sending a TokenUpdate with eventType=key
-	*/ 
-	action updateAndNotify(string key, string tokenValue, float expirationTime) {
-		CacheEntry cachedEntry := tokenCache[key];
-		cachedEntry.token := tokenValue;
-		cachedEntry.expiration := expirationTime;
-		send TokenUpdate(key, cachedEntry.token) to TokenUpdate.SUBSCRIBE_CHANNEL;	
-	}
+    action onload() {
+        log "Loaded monitor TokenFactory" at INFO;
+        spawn serveTokenRelatedRequests() to context("new_context");
+    }
+    
+    event CacheEntry {
+        string token;
+        float expiration;
+        integer consumers;
+    }
+
+    event StopEventListener {
+        string type;
+    }
+
+    dictionary<string, CacheEntry> tokenCache;
+    
+    /**
+     * Listens for all Token related requests and keep sending token refresh
+     * notifications to respective consumers using TokenUpdates
+     */
+    action serveTokenRelatedRequests() {
+        monitor.subscribe(TokenRegistrationRequest.SEND_CHANNEL);
+        monitor.subscribe(Event.SUBSCRIBE_CHANNEL);
+
+        on all TokenRegistrationRequest() as req {
+            if not tokenCache.hasKey(req.eventType) {
```

Comment:
Could put this block of code into an onTokenRegistrationRequest action?
And the same for the on RegistrationCancellationReq block as well.
What do you think?

Reply:
I don't think that is necessary. In my opinion, either way is fine. EPL users/developers should be able to understand the code perfectly when followed by **on all  TokenRegistrationRequest(..)**. Separating the logic would help clear the cluter a little bit.